### PR TITLE
Add Production Metrics in Prometheus format

### PR DESCRIPTION
**Pull request by simon-mo**
This one builds on #1662. It adds `aioprometheus` as a dependency which is very lightweight. It exposes the metrics as we perform the regular logging pass in engine step. The memory usage is very small and constant (there is no history, only current state). The metrics is designed to be scraped and stored by external service. 

We currently just add metrics in engine step. Follow up: https://github.com/vllm-project/vllm/issues/1870

Here's an example of the metrics endpoint output. 
```
# HELP exceptions_total_counter Total number of requested which generated an exception
# TYPE exceptions_total_counter counter
# HELP requests_total_counter Total number of requests received
# TYPE requests_total_counter counter
requests_total_counter{method="POST",path="/v1/chat/completions"} 1
# HELP responses_total_counter Total number of responses sent
# TYPE responses_total_counter counter
responses_total_counter{method="POST",path="/v1/chat/completions"} 1
# HELP status_codes_counter Total number of response status codes
# TYPE status_codes_counter counter
status_codes_counter{method="POST",path="/v1/chat/completions",status_code="200"} 1
# HELP vllm:avg_generation_throughput_toks_per_s Average generation throughput in tokens/s.
# TYPE vllm:avg_generation_throughput_toks_per_s gauge
vllm:avg_generation_throughput_toks_per_s{model_name="facebook/opt-125m"} 165.37719973407303
# HELP vllm:avg_prompt_throughput_toks_per_s Average prefill throughput in tokens/s.
# TYPE vllm:avg_prompt_throughput_toks_per_s gauge
vllm:avg_prompt_throughput_toks_per_s{model_name="facebook/opt-125m"} 0.0
# HELP vllm:cpu_cache_usage_perc CPU KV-cache usage. 1 means 100 percent usage.
# TYPE vllm:cpu_cache_usage_perc gauge
vllm:cpu_cache_usage_perc{model_name="facebook/opt-125m"} 0.0
# HELP vllm:gpu_cache_usage_perc GPU KV-cache usage. 1 means 100 percent usage.
# TYPE vllm:gpu_cache_usage_perc gauge
vllm:gpu_cache_usage_perc{model_name="facebook/opt-125m"} 0.002364273204903678
# HELP vllm:num_requests_running Number of requests that is currently running for inference.
# TYPE vllm:num_requests_running gauge
vllm:num_requests_running{model_name="facebook/opt-125m"} 1
# HELP vllm:num_requests_swapped Number requests swapped to CPU.
# TYPE vllm:num_requests_swapped gauge
vllm:num_requests_swapped{model_name="facebook/opt-125m"} 0
# HELP vllm:num_requests_waiting Number of requests waiting to be processed.
# TYPE vllm:num_requests_waiting gauge
vllm:num_requests_waiting{model_name="facebook/opt-125m"} 0
```

### CODE REVIEW

**Comment by ichernev** [comment]
That looks great!

Sorry for the delay

**Review by Yard1 — APPROVED** [review]
Looks good to me!

**Review by WoosukKwon — COMMENTED** [review]
@simon-mo Thanks for the awesome work! Left some minor comments.

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -639,7 +650,7 @@ def _log_system_stats(
                     f"Swapped: {len(self.scheduler.swapped)} reqs, "
                     f"Pending: {len(self.scheduler.waiting)} reqs, "
                     f"GPU KV cache usage: {gpu_cache_usage * 100:.1f}%, "
-                    f"CPU KV cache usage: {cpu_cache_usage * 100:.1f}%")
+                    f"CPU KV cache usage: {cpu_cache_usage * 100:.1f}%,")
```
[the remark this answers is not in the document]
Comment (simon-mo) [reply vllm/engine/llm_engine.py:653]: ah it's carried over from previous PR probably.

**Comment by simon-mo** [comment]
@WoosukKwon updated!

**Comment by WoosukKwon** [comment]
@simon-mo A dumb question: Can you provide an example script that I can test the PR?

**Comment by simon-mo** [comment]
```
python -m vllm.entrypoints.openai.api_server
curl http://localhost:8000/metrics
```

**Comment by WoosukKwon** [comment]
I see. I was testing `http://localhost:8000/metrics/v1/metrics` 😓 Do you think it's better to put it to `http://localhost:8000/metrics` (without v1)?

**Comment by Yard1** [comment]
It's probably not something you want to expose to users, nor is it a part of OpenAI API spec. It's better to leave it at `/metrics` IMO.

**Review by WoosukKwon — APPROVED** [review]
LGTM.

**Comment by WoosukKwon** [comment]
@simon-mo WDYT?
> I see. I was testing http://localhost:8000/v1/metrics 😓 Do you think it's better to put it to http://localhost:8000/metrics (without v1)?

**Comment by beginlner** [comment]
Hi @simon-mo , it seems like not working properly when `engine_use_ray=True`.

**Comment by ronensc** [comment]
Thank you for the PR! I noticed that [`aioprometheus`](https://github.com/claws/aioprometheus) is being used for exposing Prometheus metrics. However, it seems like there are alternative libraries that might offer additional features.

Here are a few considerations:

- **Official Prometheus Python Library:** The official library can be found at [https://github.com/prometheus/client_python](https://github.com/prometheus/client_python). It includes out-of-the-box support for various metrics, such as [GC, process, and platform](https://prometheus.github.io/client_python/collector/#disabling-default-collector-metrics) metrics.

- **Alternative Libraries:** Two alternative libraries that build on the official library and provide additional features are:
  1. [prometheus-fastapi-instrumentator](https://github.com/trallnag/prometheus-fastapi-instrumentator): Exposes FastAPI metrics.
  2. [starlette_exporter](https://github.com/stephenhillier/starlette_exporter): Exposes Starlette and FastAPI metrics. 

   Both of these libraries include HTTP-related metrics similar to `aioprometheus` but also cover latency and size metrics, which are not present in `aioprometheus`.

- **Popularity and Metrics Coverage:** It's worth noting that these alternative libraries appear to be more popular based on GitHub stars. Additionally, they provide a broader range of metrics compared to `aioprometheus`.

I'm curious to understand the rationale behind choosing `aioprometheus` over the official library or one of these alternatives. Could you please share the reasons for this decision?

**Comment by ronensc** [comment]
> Thank you for the PR! I noticed that [`aioprometheus`](https://github.com/claws/aioprometheus) is being used for exposing Prometheus metrics. However, it seems like there are alternative libraries that might offer additional features.
> 
> Here are a few considerations:
> 
>     * **Official Prometheus Python Library:** The official library can be found at https://github.com/prometheus/client_python. It includes out-of-the-box support for various metrics, such as [GC, process, and platform](https://prometheus.github.io/client_python/collector/#disabling-default-collector-metrics) metrics.
> 
>     * **Alternative Libraries:** Two alternative libraries that build on the official library and provide additional features are:
>       
>       1. [prometheus-fastapi-instrumentator](https://github.com/trallnag/prometheus-fastapi-instrumentator): Exposes FastAPI metrics.
>       2. [starlette_exporter](https://github.com/stephenhillier/starlette_exporter): Exposes Starlette and FastAPI metrics.
>       
>       Both of these libraries include HTTP-related metrics similar to `aioprometheus` but also cover latency and size metrics, which are not present in `aioprometheus`.
> 
>     * **Popularity and Metrics Coverage:** It's worth noting that these alternative libraries appear to be more popular based on GitHub stars. Additionally, they provide a broader range of metrics compared to `aioprometheus`.
> 
> 
> I'm curious to understand the rationale behind choosing `aioprometheus` over the official library or one of these alternatives. Could you please share the reasons for this decision?

@ichernev @simon-mo @WoosukKwon @Yard1 Could you please share your thoughts on this matter?

**Pull request merged**